corehq/apps/locations/bulk_management.py

```diff
@@ -0,0 +1,652 @@
+        db_object = self._as_db_object(old_collection, parent_type, domain)
+        if self.do_delete:
+            db_object.delete()
+            return None
+
+        db_object.parent_type = parent_type
+        # todo, expand_from and sync_to
+        for attr in ['name', 'code', 'shares_cases'
+                     'view_descendants']:
+            setattr(db_object, attr, getattr(self, attr, None))
+
+        if self._needs_save(old_collection):
+            db_object.save()
+
+        return db_object
+
+
+class LocationStub(object):
+    """
+    A representation of location excel row
+    """
+    titles = LOCATION_SHEET_HEADERS
+
+    def __init__(self, name, site_code, location_type, parent_code, location_id,
+                 do_delete, external_id, latitude, longitude, index):
+        self.name = name
+        self.site_code = site_code
+        self.location_type = location_type
+        self.location_id = location_id
+        self.parent_code = parent_code or 'TOP'
+        self.latitude = latitude or None
+        self.longitude = longitude or None
+        self.do_delete = do_delete
+        self.external_id = external_id
+        self.index = index
+        self.is_new = False
+        if not self.location_id and not self.site_code:
+            raise LocationExcelSheetError(
+                "Location in sheet '{}', at row '{}' doesn't contain both location_id and site_code"
+                .format(self.location_type, self.index)
+            )
+
+    @classmethod
+    def from_excel_row(cls, row, index, location_type):
+        name = row.get(cls.titles['name'])
+        site_code = row.get(cls.titles['site_code'])
+        location_type = location_type
+        location_id = row.get(cls.titles['location_id'])
+        parent_code = row.get(cls.titles['parent_code'])
+        latitude = row.get(cls.titles['latitude'])
+        longitude = row.get(cls.titles['longitude'])
+        do_delete = row.get(cls.titles['do_delete'], 'N').lower() in ['y', 'yes']
+        external_id = row.get(cls.titles['external_id'])
+        index = index
+        return cls(name, site_code, location_type, parent_code, location_id,
+                   do_delete, external_id, latitude, longitude, index)
+
+    def autoset_location_id_or_site_code(self, old_collection):
+        # if one of location_id/site_code are missing, lookup for the other in
+        # location_id/site_code pairs autoset the other if found.
+        # self.is_new is set to True if new location
+
+        if self.location_id and self.site_code:
+            return
+
+        if not self.location_id and not self.site_code:
+            # Both can't be empty, this should have already been caught in initialization
+            raise Exception(
+                "Location in sheet '{}', at row '{}' doesn't contain both location_id and site_code"
+                .format(self.location_type, self.index)
+            )
```

Comment:
This block seems fine to drop, as the properties are set in the constructor and checked there